File: Visual Studio LIghtswitch 2012/Chapter13/Chapter13/ApressExtensionVB/ApressExtensionVB/ApressExtensionVB.Common/Metadata/ModuleLoader.vb
'You can use and redistribute the code from this book (and sample application) for non-commercial and 
'most commercial purposes as long as you acknowledge the source and authorship. 
'You should note the source of the code in any documentation as well as in the program code itself (as a comment). 
'The acknowledgment should include author, title, publisher, and ISBN. 
'An example of such an acknowledgment would be "Derived from Listing 2-10, LightSwitch 2012 by Tim Leung, published by Apress, ISBN 9781430250715".

Imports System.Collections.Generic
Imports System.ComponentModel.Composition
Imports System.IO
Imports System.Resources
Imports System.Reflection

Imports Microsoft.LightSwitch.Model

<Export(GetType(IModuleDefinitionLoader))>
<ModuleDefinitionLoader("ApressExtensionVB")>
Friend Class ModuleLoader
    Implements IModuleDefinitionLoader

    Public Function GetModelResourceManager() As ResourceManager Implements IModuleDefinitionLoader.GetModelResourceManager
        Return My.Resources.ModuleResources.ResourceManager
    End Function

    Public Function LoadModelFragments() As IEnumerable(Of Stream) Implements IModuleDefinitionLoader.LoadModelFragments
        Dim assembly As Assembly = assembly.GetExecutingAssembly()
        Dim fragmentStreams As IList(Of Stream) = New List(Of Stream)

        For Each resourceName In assembly.GetManifestResourceNames()
            If resourceName.EndsWith(".lsml", StringComparison.Ordinal) Then
                fragmentStreams.Add(assembly.GetManifestResourceStream(resourceName))
            End If
        Next

        Return fragmentStreams
    End Function

End Class

File: Visual Studio LIghtswitch 2012/Chapter12/Chapter12/ApressExtensionCS/ApressExtensionCS/ApressExtensionCS.Common/Metadata/ModuleLoader.cs
//You can use and redistribute the code from this book (and sample application) for non-commercial and 
//most commercial purposes as long as you acknowledge the source and authorship. 
//You should note the source of the code in any documentation as well as in the program code itself (as a comment). 
//The acknowledgment should include author, title, publisher, and ISBN. 
//An example of such an acknowledgment would be "Derived from Listing 2-10, LightSwitch 2012 by Tim Leung, published by Apress, ISBN 9781430250715".

using System;
using System.Collections.Generic;
using System.ComponentModel.Composition;
using System.IO;
using System.Reflection;
using System.Resources;

using Microsoft.LightSwitch.Model;

using ApressExtensionCS.Resources;

namespace ApressExtensionCS.Metadata
{
    [Export(typeof(IModuleDefinitionLoader))]
    [ModuleDefinitionLoader("ApressExtensionCS")]
    internal class ModuleLoader :
        IModuleDefinitionLoader
    {
        #region IModuleDefinitionLoader Members

        ResourceManager IModuleDefinitionLoader.GetModelResourceManager()
        {
            return ModuleResources.ResourceManager;
        }

        IEnumerable<Stream> IModuleDefinitionLoader.LoadModelFragments()
        {
            Assembly assembly = Assembly.GetExecutingAssembly();
            IList<Stream> fragmentStreams = new List<Stream>();

            foreach (string resourceName in assembly.GetManifestResourceNames())
            {
                if (resourceName.EndsWith(".lsml", StringComparison.Ordinal))
                {
                    fragmentStreams.Add(assembly.GetManifestResourceStream(resourceName));
                }
            }

            return fragmentStreams;
        }

        #endregion
    }
}